examples/c/usdt.c

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+/* Copyright (c) 2022 Hengqi Chen */
+#include <signal.h>
+#include <unistd.h>
+#include <setjmp.h>
+#include <linux/limits.h>
+#include "usdt.skel.h"
+
+static volatile sig_atomic_t exiting;
+static const char *usdt_provider = "libc";
+static const char *usdt_name = "setjmp";
+static jmp_buf env;
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	return vfprintf(stderr, format, args);
+}
+
+static int get_libc_path(char *path)
+{
+	FILE *f;
+	char buf[PATH_MAX] = {};
+	char *filename;
+	float version;
+
+	f = fopen("/proc/self/maps", "r");
+	if (!f)
+		return -errno;
+
+	while (fscanf(f, "%*x-%*x %*s %*s %*s %*s %[^\n]\n", buf) != EOF) {
+		if (strchr(buf, '/') != buf)
+			continue;
+		filename = strrchr(buf, '/') + 1;
+		if (sscanf(filename, "libc-%f.so", &version) == 1) {
+			memcpy(path, buf, strlen(buf));
+			fclose(f);
+			return 0;
+		}
+	}
+
+	fclose(f);
+	return -1;
+}
+
+void usdt_trigger() {
+	setjmp(env);
+}
+
+int main(int argc, char **argv)
+{
+	char libc_path[PATH_MAX] = {};
+	struct usdt_bpf *skel;
+	int err;
+
+	libbpf_set_strict_mode(LIBBPF_STRICT_ALL);
+	libbpf_set_print(libbpf_print_fn);
+
+	skel = usdt_bpf__open_and_load();
+	if (!skel) {
+		fprintf(stderr, "Failed to open/load BPF skeleton\n");
+		return 1;
+	}
+
+	err = get_libc_path(libc_path);
+	if (err) {
+		fprintf(stderr, "Failed to get libc path\n");
+		goto cleanup;
+	}
+
+	skel->links.usdt_test = bpf_program__attach_usdt(skel->progs.usdt_test, getpid(),
```

Comment:
the way you specified BPF_USDT() program's SEC() it should be auto-attachable, so you can use iether usdt_bpf__attach() for skeleton or bpf_program__attach() for one specific program. But it would be good to demonstrate both auto-attach capabilities and have an example of explicit attach like you do here. So maybe let's have two BPF programs (I think it's fine if it's for the same USDT probe, just let's have two different bpf_printk() messages).

But either way you shouldn't need to specify absolute path for libc (though do test this, of course). Thanks!